ENH add hash based unique

adrinjalali opened the pull request:
This calculates unique values with an unordered hashset in C++ for certain dtypes.

Towards https://github.com/numpy/numpy/issues/11136

adrinjalali committed:
ENH add hash based unique

adrinjalali committed:
getting closer

ngoldbaum commented:
I marked this as draft in the github UI, hope you don't mind.

adrinjalali commented:
Thanks, I had intended to have it as draft, but forgot.

adrinjalali committed:
trying to expose as a module

adrinjalali committed:
trying to create a module

adrinjalali committed:
fix build

adrinjalali committed:
...

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into unique-cpp

adrinjalali committed:
segfault fix, imported numpy

adrinjalali committed:
getting unique back, refcount issues exist

adrinjalali committed:
unique works

adrinjalali committed:
remove header

adrinjalali commented:
Still got work to do regarding figuring out dtypes / type lengths supported on all platforms and all, and figuring out a way to filter dtypes that should be included in the hash based unique, but for now, this is giving me a somewhat 10x speedup on 100_000_000 to 1_000_000_000 data points.

adrinjalali committed:
cleanups and comments

adrinjalali committed:
change type

adrinjalali commented:
Example times on 1,000,000,000 samples:

```
unique count (hashmap):  1000
7.815 secs
unique count (numpy main):  1000
21.436 secs
unique count (polars):  1000
16.768 secs
```

Now that the code is pretty small, is this something we'd like to have in numpy?

cc @seberg , @rgommers since you've been involved in this.

I still need to fix the compiler issues for all different compilers, and maybe figure out the macro defs to add 96 and 128 bit data, but for now I think we can discuss if this is the right thing to do.

Also, the script I use to benchmark:

```py
import time
import numpy as np
from numpy._core.unique import unique_hash
import polars as pl

arr = np.random.randint(0, 1000, 1000_000_000)

time_start = time.perf_counter()
print("unique count (hashmap): ", len(unique_hash(arr)))
time_elapsed = (time.perf_counter() - time_start)
print ("%5.3f secs" % (time_elapsed))

time_start = time.perf_counter()
print("unique count (numpy main): ", len(np.unique(arr)))
time_elapsed = (time.perf_counter() - time_start)
print ("%5.3f secs" % (time_elapsed))

time_start = time.perf_counter()
print("unique count (polars): ", len(pl.Series(arr).unique()))
time_elapsed = (time.perf_counter() - time_start)
print ("%5.3f secs" % (time_elapsed))
```

Haven't included pandas since I guess they still need to do a release with numpy2 first.

ngoldbaum commented:
This seems like a reasonable amount of code to maintain; it's more or less a wrapper around the C++ standard library `unordered_map`.

ngoldbaum commented:
Also totally a style thing, but to keep things consistent it probably makes more sense to add the `unique_hash` function to the main `numpy._core._multiarray_umath` module that most other numpy core functionality lives in than to define a new `np._core.unique` module.

rgommers commented:
> Now that the code is pretty small, is this something we'd like to have in numpy?

Nice work. I agree with what @ngoldbaum said above. This seems like a very clean and concise implementation.

adrinjalali committed:
fix for initialization issue

adrinjalali committed:
trying to move module

seberg commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,168 @@
+    {sizeof(npy_uint16), unique<npy_uint16>},
+    {sizeof(npy_uint32), unique<npy_uint32>},
+    {sizeof(npy_uint64), unique<npy_uint64>},
+};
```
If you plug this into the main `unique` function, you should find some errors.  The point being: Maybe best not to try to be smart (bools are also a slight problem).

For floats the hash function has to account for `0` and `-0` being different (the NaN behavior may be trickier, since I am not sure C++ hashmap accounts for it, even if the float hashing does, does the comparison use something NaN aware?).

This all seems light-weight enough that I am happy to go there (not sure if unordered map is ideal, but probably good enough in practice; for one I am not sure if unordered map actually does hash randomization).

I am always a bit concerned that I would like to be able to provide a public API for others to implement this for their DTypes, and one that ideally doesn't use any NumPy array at all.  I think that would mean splitting this into 3 functions:
* One function which sets up the iterator and does the outer iteration (the `iternext`).  It then would call, unfortunately maybe 3 other functions (although that can all just be templated for now I think):
  * Create a new hash-map of the right type. 
  * Call one function which just deals with updating the hash-map for the 1-D slice (i.e. that `while (count--)` loop).
  * Split out the finalization of the hash-map into a NumPy array (or a single contiguous allocation).

If we had those, it should be straight forward to make them available to user DTypes.  Of course such a reorganization isn't terrible hard even if we don't do it now.

---

This code also needs handling with C++ exceptions (at the very least memory errors), so that is a bit of a problem, we may need to steal a generic translator (e.g. from `pybind11`), so that you can do `translate_errors_cpp_meth<PyArray_Unique>` when registering below.

seberg replied:
Actually, strike that: There is a bigger problem with C++ exceptions: You would have to use a helper `OwnedRef` object in all such code to ensure you clean up correctly.
So catching is only straight forward if no `PyObject` or manual memory handling is happening inside the functions being caught.

jorisvandenbossche commented:
> Haven't included pandas since I guess they still need to do a release with numpy2 first.

You could install the nightly wheels from the scientific python channel if you want (I don't think there is any (compiled) library that already has a 2.0 release?)

I tried your benchmark case locally with pandas. I didn't fetch this branch to compare with, but also ran with released numpy as a point of comparison:

```
%time len(pd.unique(arr))
CPU times: user 2.74 s, sys: 0 ns, total: 2.74 s
Wall time: 2.74 s

%time len(np.unique(arr))
CPU times: user 39.7 s, sys: 1.15 s, total: 40.8 s
Wall time: 40.9 s
```

Related to `std::unordered_map`, you find quite some content "on the internet" complaining about how slow it is compared to other implementations (I remember from looking a bit around some time ago for pandas, wondering if we should consider using a different implementations; pandas uses khash). But a lost of those posts are also quite some years old, and C++ compilers might have improved their implementation. It's definitely convenient to use)

adrinjalali commented:
@ngoldbaum 
> Also totally a style thing, but to keep things consistent it probably makes more sense to add the `unique_hash` function to the main `numpy._core._multiarray_umath` module that most other numpy core functionality lives in than to define a new `np._core.unique` module.

I tried and failed. One is C++ the other C, and they also have different macros related to compilation. So not sure how to merge them.

@seberg 

- bools and floats: good points. Will make this PR "less smart" and dtype specific code for those types. Although I think those can also be a separate PR to keep this one small.
- refactoring: I'll try
- Could you please elaborate on `OwnedRef` please? I'll checkout the pybind thingy to see how exceptions should be handled in the meantime.
- randomized hash: they're compiler specific and AFAIK at least for gcc they might be anything by randomized. Pandas uses https://github.com/veorq/SipHash which is certainly cryptographically secure. But there are other randomized hash functions which are faster, but not as secure. We could also consider adding that / changing to it in a separate PR maybe? Or would you say having a terrible hash function as the one in this PR is too bad to be merged into numpy?

@jorisvandenbossche 
Yes, pandas is really fast, and last I compared, for unique, it was faster than polars. But it's also a huge piece of code on the pandas side. Porting that in here was the first thing I tried (https://github.com/numpy/numpy/pull/25596).

seberg commented:
> I tried and failed. One is C++ the other C

Should work, I think but I guess you could defer that for now.  You will have to export an `extern "C" {` function to be called from multiarray-module.

> Could you please elaborate on `OwnedRef` please?

In C, you do cleanup with a `goto error: /* cleanup code */`.  In C++ you don't use `goto error:` but rely on exceptions, when an exception gets triggered (which could be in many places e.g. when out of memory), you need to run that cleanup code.
In particular, if you have `PyObject *` you must decref them.  There are three solutions to that:
1. You use C++ only in a core function that uses only borrowed references.
2. You wrap everything in a `try/catch`...
3. You introduce a new `OwnedRef` which which does nothing but store a `PyObject *` but defines a dealloc.  That way C++ does the right thing.

In this case, I am not sure what is better, it might be 1 (have a simple C entry-point and just call into some C++ code that maybe doesn't even need objects at all) or 3 if you want to work with objects in C++.

> having a terrible hash function as the one in this PR is too bad to be merged into numpy?

I doubt it matters much, there is nothing cryptographic here, the issue would be denial of service by creating a dataset with a huge number of unique values which all hash to the same thing.  Note that e.g. Python only randomizes the hash for strings (unless dicts randomize them for non-strings?).

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into unique-cpp

adrinjalali committed:
Revert "trying to move module"

This reverts commit 3db33498f5c4df90ebb4f4a8ccf11fa2d72ff574.

JensWehner commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,168 @@
+    NpyIter_IterNextFunc *iternext;
+    char** dataptr;
+    npy_intp* strideptr,* innersizeptr;
+    std::unordered_map<T, char> hashmap;
```
Just curious, would an unordered_set not be enough?

adrinjalali committed:
use unordered_set and use a finally construct to handle exceptions

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into unique-cpp

adrinjalali committed:
handle C++ exceptions, and use explicit types

adrinjalali commented:
Still some work to do. But I've switched to explicit type checking, and only including integers for now. Also, using a nice C++ hack to handle deallocation / DECREF stuff when exceptions happen.

For now at least we like the performance:

```py
import time
import numpy as np
from numpy._core.unique import unique_hash
import polars as pl
import pandas as pd

arr = np.random.randint(0, 1000, 1000_000_000)

time_start = time.perf_counter()
print("unique count (hashmap): ", len(unique_hash(arr)))
time_elapsed = (time.perf_counter() - time_start)
print ("%5.3f secs" % (time_elapsed))

time_start = time.perf_counter()
print("unique count (numpy, using hashmap): ", len(np.unique(arr, sorted=False)))
time_elapsed = (time.perf_counter() - time_start)
print ("%5.3f secs" % (time_elapsed))

time_start = time.perf_counter()
print("unique count (numpy, return_counts): ", len(np.unique(arr, return_counts=True)[0]))
time_elapsed = (time.perf_counter() - time_start)
print ("%5.3f secs" % (time_elapsed))

time_start = time.perf_counter()
print("unique count (polars): ", len(pl.Series(arr).unique()))
time_elapsed = (time.perf_counter() - time_start)
print ("%5.3f secs" % (time_elapsed))

time_start = time.perf_counter()
print("unique count (pandas): ", len(pd.Series(arr).unique()))
time_elapsed = (time.perf_counter() - time_start)
print ("%5.3f secs" % (time_elapsed))


unique count (hashmap):  1000
1.464 secs
unique count (numpy, using hashmap):  1000
2.283 secs
unique count (numpy, return_counts):  1000
29.750 secs
unique count (polars):  1000
4.610 secs
unique count (pandas):  1000
2.334 secs
``` 

That's quite a significant amount of overhead when calling this unique function via the numpy machinery though. Something to look at.

I'll look at the tests next.

ngoldbaum commented:
I believe there's at least one stalled PR related to a hash-based unique implementation, which would fix that IIUC. I don't remember the details about why that's held up but if you're up for some github archeology that will likely be enlightening :)

adrinjalali commented:
You mean a stalled PR other than this one? 😅 I tried, but didn't find anything.

ngoldbaum commented:
Sorry I'm misremembering, please ignore me.

adrinjalali committed:
make it C importable

adrinjalali committed:
add missing header file

adrinjalali committed:
fix skip API test

adrinjalali committed:
rename _core.unique to _core._unique

adrinjalali committed:
use _unique name

adrinjalali commented on numpy/lib/_arraysetops_impl.py:
```diff
@@ -335,21 +341,40 @@ def reshape_uniq(uniq):
+
+    # masked arrays are not supported yet.
+    if not optional_indices and not return_counts and not np.ma.is_masked(ar):
+        # from numpy._core._multiarray_umath import unique_hash
+        from numpy._core._unique import unique_hash
```
@ngoldbaum switching the commented status of these two lines to import from `_multiarray_umath` gives a segfault at this line:

https://github.com/numpy/numpy/pull/26018/files#diff-32d1424a27bdc13b58b84b0f68013d5ea1fc5bcf48ad954066b437431e3d5939R138

Not sure why, I could use some hints about this one.

adrinjalali commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,189 @@
+    } while(iternext(iter));
+
+    // then we iterate through the map's keys to get the unique values
+    T* res = new T[hashset.size()];
```
does this need any refcounting?

seberg replied:
What we need is that we have to make sure two things:
1. We need to ensure we clean up `iter` (call `NpyIter_Dealloc`) even if there are C++ exceptions.  `new` can raise such an exception (but I am sure many things here can).
2. If a C++ exception occurs, we need to convert it into some reasonable Python error.  In most cases these should be critical errors (i.e. `RuntimeError` is OK I think), although distinguishing `MemoryError` may be good.
   We _might_ have a pattern for this in the fft ufunc code.

Otherwise, I think I can live with this.  We could move this `unique` function onto the dtype.
A question for that would be whether we can refactor the function so that it actually becomes:
```
dtype.get_unique(&update_function, &finalize_function, &auxdata)

do {
     if (update_function(context, data, stride, count, auxdata) < 0) {
          break;  /* ouch, error */
     }
} while (iternext)

void *res = finalize_function(context, auxdata);
```
I _think_ that is not a problem (and I suspect required to disect out knowing about NumPy arrays and creating a stable C-API).

TL;DR:  Should we meet up and walk through the code once (if just to force me to  look at it closer?).  I think I am fine with this, although I might want to at least put a hint to what I wrote above into the code to tell people how we may achieve that users can implement unique for their dtypes.

(I think I might like to already split it into that 1-D `update_func` and a `finalize_func`, just because that is such a typical pattern in NumPy to organize things around 1-D loops.)

adrinjalali replied:
There shouldn't be any issues with iter deallocation with the current code.

Sorry for the delay, end of year release and holidays took some time. I'd be very happy to have a call and go through this if you're available next week.

seberg replied:
About the exceptions, have used this before: https://github.com/numpy/numpy/blob/dc78e309dd96846dc345af85017270862c642f70/numpy/fft/_pocketfft_umath.cpp#L32-L51  (this is a template trick, but maybe you know a better way).

If we are sure that the only exceptions possible can come from `new` (which seems pretty likely, but not sure!) then you can also use `new(std::nothrow)`, though.

adrinjalali reviewed (commented):
Update:

This code now passes usual tests. It fails the freethreaded CI with this error:

```py
  >           from numpy._core._unique import unique_hash
  E           RuntimeWarning: The global interpreter lock (GIL) has been enabled to load module 'numpy._core._unique', which has not declared that it can run safely without the GIL. To override this behavior and keep the GIL disabled (at your own risk), run with PYTHON_GIL=0 or -Xgil=0.
```

I'm not sure how to fix that (haven't touched nogil stuff yet).

@seberg I think I've addressed your concerns, would you mind having a look?

I think we can leave the following for followup PRs:
- Add support for string types
- Add support for masked arrays
- Add support for `return_counts`

The first two are used in our use case in scikit-learn.

ngoldbaum commented:
> I'm not sure how to fix that (haven't touched nogil stuff yet).

The new extension module you added needs to be thread safe. I don't see any obvious thread safety issues and I would assume `std::unordered_set` is safe, but no idea if there are platforms where that isn't the case.

Assuming it is safe, the new extension module you added needs to set the `Py_MOD_GIL` slot. See https://py-free-threading.github.io/porting/#__tabbed_1_1 for examples and also grep the numpy codebase for `Py_MOD_GIL`.

seberg commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,189 @@
+{
+    PyArray_ImportNumPyAPI();
+    return PyModule_Create(&uniquemodule);
+}
```
I feel like we should be able to just include this into `multiarray.c` and add the private helper function there (avoiding the additional module)?

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into unique-cpp

adrinjalali committed:
add freethreaded slot

seberg reviewed (commented):
As discussed, a couple of comments (most already talked about, so don't hesitate to mark them as done).
I'll note the small thing (maybe noted elsewhere, we said: return `NotImplemented` as singleton for non-supported dtype and of course never return None on error).

Overall, I think we at least know (rough!) path that could be used to support user dtypes, so I am fine with going forward (i.e. the likely chance it won't happen for years seems OK).

adrinjalali committed:
apply comments from review

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into unique-cpp

adrinjalali commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,198 @@
+    */
+    PyArrayObject *self = NULL;
+    PyObject *res = NULL;
+    if (!PyArg_ParseTuple(args, "O&", PyArray_Converter, &self))
```
needed this

```C
#define _MULTIARRAYMODULE
```

didn't end up having much to do being its own module I think.

seberg replied:
If you define that, the function is linked in, if you don't, the function expands to a table lookup (and that table can be uninitialized).  So yeah, it works, but may be weird to do that unless you are in the same compilation unit :).

adrinjalali committed:
remove own module, fix segfault

adrinjalali committed:
raise NotImplementedError instead of returning None

adrinjalali reviewed (commented):
@seberg I think this is good for another review. As I understand, other than the `-fexceptions` compiler flag question, everything is handled.

seberg commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,190 @@
+            // Grab GIL before raising an exception
+            NPY_ALLOW_C_API;
+            PyErr_SetString(PyExc_NotImplementedError, "Unsupported dtype");
+            NPY_DISABLE_C_API;
```
Here, we always have the GIL.  You'll need to add a `finally` function/destructor to the caller where you release the GIL (similar to the NpyIter cleanup, or including in that).

I realize that this comes from the ufunc code which is in a different situation here.  I am not sure that you are already releasing the GIL, but if you are not, you could do so before the `do {} while ()` loop and grab it again after).

Just a brief comment because I happened to look at it.  Will look at it more later (ping me in a while if I forgot).

adrinjalali replied:
Now I explicitly release the GIL and grab it back in the compute heavy part.

adrinjalali committed:
release and regrab GIL

adrinjalali committed:
fix GIL issues and compile separately

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into unique-cpp

adrinjalali commented on numpy/_core/meson.build:
```diff
@@ -690,7 +690,8 @@ cpp_args_common = c_args_common + [
   cpp_args_common += [
-    '-fno-exceptions',  # no exception support
+    # unique code handles C++ exceptions
+    # '-fno-exceptions',  # no exception support
```
There's now a separate static lib build for it with the right flags.

adrinjalali reviewed (commented):
I've separated the build for this file so that we can pass a different `-fexceptions` flag to it, and not the others, and it's fine in most builds. However, it fails in a few of them with errors that I don't understand, for instance:

```sh
Missing dep: numpy/_core/libunique_hash.a.p/src_multiarray_unique.cpp.o uses numpy/_core/__multiarray_api.h (generated by CUSTOM_COMMAND)
Processed 526 nodes.
Error: There are 1 missing dependency paths.
1 targets had depfile dependencies on 1 distinct generated inputs (from 1 rules)  without a non-depfile dep path to the generator.
There might be build flakiness if any of the targets listed above are built alone, or not late enough, in a clean output directory.
```

or in windows:

```sh
  FAILED: numpy/_core/libunique_hash.a.p/src_multiarray_unique.cpp.obj
  "cl" "-Inumpy\_core\libunique_hash.a.p" "-Inumpy\_core" "-I..\numpy\_core" "-Inumpy\_core\include" "-I..\numpy\_core\include" "-I..\numpy\_core\src\common" "-IC:\hostedtoolcache\windows\Python\3.11.9\x86\Include" "-ID:\a\numpy\numpy\.mesonpy-l4odngo_\meson_cpu" "-DNDEBUG" "/MD" "/nologo" "/showIncludes" "/utf-8" "/Zc:__cplusplus" "/W2" "/EHsc" "/std:c++17" "/permissive-" "/O2" "/Gw" "/arch:SSE2" "-DNPY_HAVE_SSE" "-DNPY_HAVE_SSE2" "-DNPY_INTERNAL_BUILD" "-DHAVE_NPY_CONFIG_H" "-D_FILE_OFFSET_BITS=64" "-D_LARGEFILE_SOURCE=1" "-D_LARGEFILE64_SOURCE=1" "/Fdnumpy\_core\libunique_hash.a.p\src_multiarray_unique.cpp.pdb" /Fonumpy/_core/libunique_hash.a.p/src_multiarray_unique.cpp.obj "/c" ../numpy/_core/src/multiarray/unique.cpp
  D:\a\numpy\numpy\numpy\_core\include\numpy\ndarrayobject.h(22): fatal error C1083: Cannot open include file: '__multiarray_api.h': No such file or directory
  [51/529] Generating 'numpy\\_core\\_multiarray_umath.cp311-win32.pyd.p\\lowlevel_strided_loops.c'
```

I'd need help to fix these.

ngoldbaum commented on numpy/_core/src/multiarray/unique.h:
```diff
@@ -0,0 +1,16 @@
+#ifndef NUMPY_CORE_INCLUDE_NUMPY_MULTIARRAY_UNIQUE_H_
+#define NUMPY_CORE_INCLUDE_NUMPY_MULTIARRAY_UNIQUE_H_
+
+#include <Python.h>
```
I don't think this should be in a header - we keep all includes in the object files

ngoldbaum reviewed (commented):
EDIT: sorry what I originally wrote here is completely wrong

ngoldbaum commented on numpy/_core/meson.build:
```diff
@@ -1221,6 +1231,20 @@ if use_svml
+    'src/common',
+  ],
+  dependencies: [
+    py_dep,
```
oh actually it might be because you're missing a dependency here?

by the way, why is it so important for this to be its own module? you'd probably have an easier time just adding a new thing to the multiarray module....

seberg replied:
@ngoldbaum you probably know the meson setup well: is there a nice way to override the exception flags for this file?
Before we compile it as it's own module, maybe it is best to just compile `multiarraymodule` with exception support.
(At least my understanding is that @adrinjalali went back to this setup because it allows to keep the multiarray exception compile flags while the use of the hash-map makes it at least inconvenient, maybe impossible, for this file.)

adrinjalali replied:
The previous setup was that `unique_hash` was a separate Python module, but now it isn't and it _is_ indeed in `multiarraymodule`. However, the file needs to be compiled with different flags than the other files in that module, hence a separate `static_library` (instead of a separate `py.extension_module`) just to generate the `.so` file to then link the rest with it.

adrinjalali committed:
add np_core_dep dependency, hoping it fixes the issue

adrinjalali committed:
remove include

adrinjalali commented:
So the only failing test now is "Linux tests / debug", which uses `python-dbg`. I haven't had to run tests against that ever, so I'll try to see what's going on. I'll need to replicate this locally to see if I can reproduce, or find a way to download the coredump from the CI somehow.

```yaml
  debug:
    needs: [smoke_test]
    runs-on: ubuntu-24.04
    if: github.event_name != 'push'
    steps:
    - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
      with:
        submodules: recursive
        fetch-tags: true
        persist-credentials: false
    - name: Install debug Python
      run: |
        sudo apt-get update
        sudo apt-get install python3-dbg ninja-build
    - name: Build NumPy and install into venv
      run: |
        python3-dbg -m venv venv
        source venv/bin/activate
        pip install -U pip
        pip install . -v -Csetup-args=-Dbuildtype=debug -Csetup-args=-Dallow-noblas=true
    - name: Install test dependencies
      run: |
        source venv/bin/activate
        pip install -r requirements/test_requirements.txt
    - name: Run test suite
      run: |
        source venv/bin/activate
        cd tools
        pytest --timeout=600 --durations=10 --pyargs numpy -m "not slow"
```

adrinjalali committed:
debug ...

adrinjalali committed:
debug ...

adrinjalali committed:
Py_INCREF needs the GIL

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into unique-cpp

adrinjalali committed:
revert debug info in CI

adrinjalali commented:
@ngoldbaum @seberg this now has a green CI and I think I've addressed all the issues in the review so far. Ready for another look. I'd open separate issues for the future work we have in mind afterwards if we want a place to record them.

ngoldbaum commented:
First, sorry for rehashing the exceptions stuff you and Sebastian already worked out and thanks for clarifying. Also thank you so much for working through some C++ teething pains with us.

> this now has a green CI and I think I've addressed all the issues in the review so far

I agree - I'm pretty sure the CI failures are unrelated. I'll open an issue about them.

> I'd open separate issues for the future work we have in mind afterwards if we want a place to record them.

Opening issues to have a little discussion before doing a PR can help avoid unnecessary work so I think it's worth doing. Maybe open a tracking issue for improvements to unique and then create sub-issues that link to the tracking issue? But organize it however is most convenient to you.

ngoldbaum commented:
I'll try to do a review pass in the next few days if someone else doesn't get to it in the meantime.

seberg committed:
Merge highway submodule changes from main

seberg commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,199 @@
+                self,
+                NPY_ANYORDER,
+                NULL, // descr (use prototype's descr)
+                1 // subok
```
```suggestion
                0 // subok (function always returns base-class)
```
I don't think it matters, but noticed the discrepancy.

seberg commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,199 @@
+        auto type = PyArray_TYPE(self);
+        // we only support data types present in our unique_funcs map
+        if (unique_funcs.find(type) == unique_funcs.end()) {
+            // Grab GIL before raising an exception
+            NPY_ALLOW_C_API;
+            PyErr_SetString(PyExc_NotImplementedError, "Unsupported dtype");
+            NPY_DISABLE_C_API;
+            return NULL;
```
```suggestion
            return Py_RETURN_NOTIMPLEMENTED;
```
slight misunderstanding: Since this is an internal functionality, I think returning `NotImplemented` is the clearest, not raising the error :).  (raising an error can a bit slow, although it may have improved)

seberg commented on numpy/_core/src/multiarray/multiarraymodule.c:
```diff
@@ -4586,6 +4588,8 @@ static struct PyMethodDef array_module_methods[] = {
         "Give a warning on reload and big warning in sub-interpreters."},
     {"from_dlpack", (PyCFunction)from_dlpack,
         METH_FASTCALL | METH_KEYWORDS, NULL},
+    {"unique_hash",  (PyCFunction)array_unique,
```
NIT:
```suggestion
    {"_unique_hash",  (PyCFunction)array__unique_hash,
```
(or not the double `__`.  I am fine either way, but I think it is good to call it `unique_hash` also internally for clarity.)

seberg commented on numpy/_core/src/multiarray/multiarraymodule.c:
```diff
@@ -4586,6 +4588,8 @@ static struct PyMethodDef array_module_methods[] = {
     {"from_dlpack", (PyCFunction)from_dlpack,
         METH_FASTCALL | METH_KEYWORDS, NULL},
+    {"unique_hash",  (PyCFunction)array_unique,
+        METH_VARARGS, "Collect unique values via a hash map."},
```
NIT (if you like): Could even use `METH_O` (but changes the parsing a bit).

seberg commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,199 @@
+}
+
+template<typename T>
+PyObject* unique(PyArrayObject *self)
```
NIT:
```suggestion
static PyObject*
unique(PyArrayObject *self)
```
(still trying to use `static` I guess.  We should change the default linking on that, I am not sure we already did...)

seberg commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,199 @@
+    // or w/o an exception
+    auto iter_dealloc = finally([&]() { NpyIter_Deallocate(iter); });
+    if (iter == NULL) {
+        return Py_None;
```
```suggestion
        return NULL;
```

seberg commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,199 @@
+
+    iternext = NpyIter_GetIterNext(iter, NULL);
+    if (iternext == NULL) {
+        return Py_None;
```
```suggestion
        return NULL;
```

seberg commented on numpy/lib/_arraysetops_impl.py:
```diff
@@ -350,6 +358,31 @@ def _unique1d(ar, return_index=False, return_inverse=False,
+    if (optional_indices or return_counts) and not sorted:
+        raise ValueError(
+            "`sorted` can only be False if `return_index`, `return_inverse`, "
+            "and `return_counts` are all False."
```
Should we add that we may add support in the future (e.g. for `return_counts`)?  Or just the fun word "currently" :)
(ValueError or TypeError seems reasonable to me, though)

seberg commented on numpy/lib/_arraysetops_impl.py:
```diff
@@ -350,6 +358,31 @@ def _unique1d(ar, return_index=False, return_inverse=False,
+                hash_unique.sort()
+            # We wrap the result back in case it was a subclass of numpy.ndarray.
+            return (conv.wrap(hash_unique),)
+        except NotImplementedError:
```
This would become an `if hash_unique is not NotImplemented: ...` now.

seberg commented on numpy/lib/_arraysetops_impl.py:
```diff
@@ -182,6 +184,9 @@ def unique(ar, return_index=False, return_inverse=False,
         .. versionadded:: 1.24
 
+    sorted : bool, optional
+        If True, the unique elements are sorted.
```
```suggestion
        If True, the unique elements are sorted.

        .. versionadded:: 2.3
```

seberg commented on numpy/lib/_arraysetops_impl.py:
```diff
@@ -350,6 +358,31 @@ def _unique1d(ar, return_index=False, return_inverse=False,
+
+    # masked arrays are not supported yet.
+    if not optional_indices and not return_counts and not np.ma.is_masked(ar):
+        from numpy._core._multiarray_umath import unique_hash
```
No reason to not have this a top-level import, I think?

seberg commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,199 @@
+    {NPY_UINT8, unique<npy_uint8>},
+    {NPY_UINT16, unique<npy_uint16>},
+    {NPY_UINT32, unique<npy_uint32>},
+    {NPY_UINT64, unique<npy_uint64>},
```
You could actually add datetime by using `int64/uint64` as the function, since the NaT values don't matter for "unique".

seberg replied:
A note for other reviewers: Floats are currently left out here to avoid having to deal with a custom hash/equal implementation (due to NaN values).

seberg reviewed (commented):
I had pushed a commit to update highway to the one in `main`, but while the tests now pass, something still seems off?
So you may want to just ignore that and see what is wrong... (if it isn't just github being confused...)


Thanks for the iteration, it great to see most of the things fixed and this now being part of the main module :)!

A few comments, many of which I am happy if you ignore.  I suspect we'll improve on some of the C++ patterns over time but that is OK (I think this is the first C++ function that is directly exposed as Python function).

I think it would be good if Nathan has a brief look at this as well.  (FWIW, I am not even sure if `unordered_set` is the right choice here in the long term.)

---

This should have a release note IMO.

ngoldbaum commented:
> something still seems off?

The submodule is correct now. It shows a diff on the PR because the PR is based on an old `main` version. If @adrinjalali merges with or rebases the PR on `main` the diff will look clean again.

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into unique-cpp

adrinjalali committed:
reviews

adrinjalali committed:
add test for ValueError

adrinjalali committed:
changelog

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into unique-cpp

adrinjalali reviewed (commented):
I think I've addressed all comments.

adrinjalali committed:
use macro to return notimplemented

adrinjalali commented:
Doesn't seem like the failure is related to this PR

```py
=================================== FAILURES ===================================
___________ TestStringDiscovery.test_nested_arrays_stringlength[1.2] ___________

self = <test_array_coercion.TestStringDiscovery object at 0x3cd6fb17d010>
obj = 1.2

    @pytest.mark.parametrize("obj",
            [object(), 1.2, 10**43, None, "string"],
            ids=["object", "1.2", "10**43", "None", "string"])
    def test_nested_arrays_stringlength(self, obj):
        length = len(str(obj))
        expected = np.dtype(f"S{length}")
        arr = np.array(obj, dtype="O")
>       assert np.array([arr, arr], dtype="S").dtype == expected
E       RuntimeWarning: invalid value encountered in cast

arr        = array(1.2, dtype=object)
expected   = dtype('S3')
length     = 3
obj        = 1.2
self       = <test_array_coercion.TestStringDiscovery object at 0x3cd6fb17d010>
```

ngoldbaum commented:
The fix for that was just merged

adrinjalali committed:
Merge remote-tracking branch 'upstream/main' into unique-cpp

seberg committed:
Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

seberg committed:
MAINT,ENH: Smaller reorgs/maint and use `sorted=False` for `unique_values`

seberg commented on numpy/_core/src/multiarray/unique.cpp:
```diff
@@ -0,0 +1,180 @@
+    NPY_ALLOW_C_API;
+    PyArray_Descr *descr = PyArray_DESCR(self);
+    Py_INCREF(descr);
+    PyObject *res_obj = PyArray_NewFromDescr(
```
If there was a C++ exception after this, we would leak it... But, I don't think that is possible, so let's not worry about it.

seberg reviewed (approved):
I added a small commit to:
* Use `sorted=False` for `unique_values` (and a release note)
* Some smaller maintenance, partially just small preferences, though.
* I removed the 0-size special case.  I think it is just nice to not need it (and I don't consider 0-size to be important to optimize)

This LGTM, would be nice if someone could have a quick look over my changes.

adrinjalali commented:
There's a segfault and the docs need updating since output is not sorted now @seberg

seberg committed:
Ensure we don't iterate if iterator is empty (also change thread state slightly)

seberg commented:
> There's a segfault and the docs need updating since output is not sorted now

Whoops, need to not try to iterate (I think the setup is OK though).

The output being sorted is not documented for `unique_values`, which is the one I changed.

seberg committed:
DOC: unique_values doc examples may have different order now

(let's pretend it may not even be stable!)

seberg commented:
Just to note the observation:  The s390x _build_ segfaulted once here.

seberg commented:
Let's give this a shot, thanks @adrinjalali, we discussed it briefly yesterday and I don't think it is helpful to try to iterate here more.

There are a lot of smaller and larger follow-ups (larger e.g. thinking about using another hash-map, maybe even one with some randomization, etc.).

adrinjalali commented:
Thank y'all for all the patience, all the reviews, and all the help. This was a steep learning curve for me, doing C++ after about 12 years, and never having had done cpython bindings using its barebone API 😅 

I loved every bit of this experience, and look forward to more contributions on this little corner ❤️
